scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1643 @@
+            Number of samples to generate in the parameter space. Default is 1.
+
+        Returns
+        -------
+        sample : array_like (n, d)
+            Sobol' sample.
+
+        """
+        sample = np.empty((n, self.d), dtype=float)
+
+        if self.num_generated == 0:
+            # verify n is 2**n
+            if not (n & (n - 1) == 0):
+                warnings.warn("The balance properties of Sobol' points require"
+                              " n to be a power of 2.")
+
+            if n == 1:
+                sample = self._first_point
+            else:
+                _draw(n - 1, self.num_generated, self.d, self._sv,
+                      self._quasi, sample)
+                sample = np.concatenate([self._first_point, sample])[:n]
+        else:
+            _draw(n, self.num_generated - 1, self.d, self._sv,
+                  self._quasi, sample)
+
+        self.num_generated += n
+        return sample
+
+    def random_base2(self, m):
+        """Draw point(s) from the Sobol' sequence.
+
+        This function draws :math:`n=2^m` points in the parameter space
+        ensuring the balance properties of the sequence.
+
+        Parameters
+        ----------
+        m : int
+            Logarithm in base 2 of the number of samples; i.e., n = 2^m.
+
+        Returns
+        -------
+        sample : array_like (n, d)
+            Sobol' sample.
+
+        """
+        n = 2 ** m
+
+        total_n = self.num_generated + n
+        if not (total_n & (total_n - 1) == 0):
+            raise ValueError("The balance properties of Sobol' points require "
+                             "n to be a power of 2. {0} points have been "
+                             "previously generated, then: n={0}+2**{1}={2}. "
+                             "If you still want to do this, the function "
+                             "'Sobol.random()' can be used."
+                             .format(self.num_generated, m, total_n))
+
+        return self.random(n)
+
+    def reset(self):
+        """Reset the engine to base state.
+
+        Returns
+        -------
+        engine: Sobol
+            Engine reset to its base state.
+
+        """
+        self._quasi = self._shift.copy()
+        self.num_generated = 0
+        return self
+
+    def fast_forward(self, n):
+        """Fast-forward the sequence by `n` positions.
+
+        Parameters
+        ----------
+        n: int
+            Number of points to skip in the sequence.
+
+        Returns
+        -------
+        engine: Sobol
+            The fast-forwarded engine.
+
+        """
+        if self.num_generated == 0:
+            _fast_forward(n - 1, self.num_generated, self.d,
+                          self._sv, self._quasi)
+        else:
+            _fast_forward(n, self.num_generated - 1, self.d,
+                          self._sv, self._quasi)
+        self.num_generated += n
+        return self
+
+
+class MultivariateNormalQMC(QMCEngine):
```

Comment:
I find myself considering the statements:

"A `MultivariateNormalQMC` _is a_ `QMCEngine`"

and

"A `MultivariateNormalQMC` _has a_ `QMCEngine`"

As the class is written, both are true. Should it be that way?

Reply:
both are true. `MultivariateNormalQMC` implements the `QMCEngine` API, to achieve that it wraps a `QMCEngine` as an attribute.

Reply:
I'll resolve this now and reconsider when I get to review these classes in detail, after finishing with `discrepancy`. But this still doesn't strike me as the best way to organize things. I'm not an OOP expert, and so I definitely defer to anyone who is, but it would make more sense to me if a `MultivariateNormalQMC` simply _has_ a `QMCEngine` but isn't one itself. 

Suppose `qmc` is a `QMCEngine` and `dist` is a `MultivariateNormalQMC` that uses `qmc` to generate the random data. If the user wants to reset the `QMCEngine`, they just call `qmc.reset()`, not `dist.reset()`. 

Even if does make sense for `MultivariateNormalQMC` to have a `reset` method that calls `qmc.reset()`, I don't see a need for `MultivariateNormalQMC` to _inherit_  everything from `QMCEngine`.

Suppose we did something like what is proposed in gh-13368, and we are able to provide `qmc` as the `random_state` for any SciPy stats distribution's method. Would we then want `scipy.stats.rv_generic` (or one of its subclasses) to inherit from `QMCEngine`? Should `scipy.stats.rv_generic` inherit from a `np.random.RandomState/Generator`? If not, why should `MultivariateNormalQMC`?